lib/include/platform/atomic.h

```diff
@@ -0,0 +1,294 @@
+/*
+ * Copyright (C) 2019 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file atomic.h
+ * @brief Atomic operations.
+ */
+
+#ifndef ATOMIC_H
+#define ATOMIC_H
+
+/* Standard includes. */
+#include <stdint.h>
+#include <stdbool.h>
+
+/* Check which compiler is used. */
+#include "compiler.h"
+
+/* Compiler specific implementation. */
+#if ( COMPILER_OPTION_GCC_ATOMIC_BUILTIN == 1 )
+
+/* Default GCC compiler.
+ * If GCC is used && the target architecture supports atomic instructions,
+ * (e.g. ARM, Xtensa, ... )
+ * there's no reason to not use GCC built-in atomics.
+ * Thus, no user extension is provided in this branch.
+ *
+ * If you believe atomic built-in is not compiled correctly by GCC for your target,
+ * please use the other branch and provide your own implementation for atomic.
+ *
+ * attribute flatten is not needed in this branch. 
+ */
+    #define FORCE_INLINE_FLATTEN    inline __attribute__( ( always_inline ) )
```

Comment:
A note for Lorenzo -- flatten is removed here, since with gcc built-in (default implementation), flatten does not affect the compilation result. The macro name still has "FLATTEN" in it, and user needs to make sure nested function calls are flattened and inlined.

Reply:
I perused other code bases, and it seems nobody uses flatten. I cannot find any evidence as to why though. Should we remove it for now? Seems confusing if it does not had value... we could instead add a comment on/

Reply:
Agreed. I'll updated. Comments will be left in ```atomic_user_override_template.h```, saying user needs to check inline behavior in disassembly since there's a function call. 

Limited example I can find -- intel graphic driver (SandyBridge's New Acceleration component) https://github.com/intel/iotg-lin-gfx-ddx/blob/master/src/sna/sna_glyphs.c#L583. Their [README](https://github.com/intel/iotg-lin-gfx-ddx/blob/master/src/sna/README) explained a little bit about the overall design. (I won't pretend to have a full understanding of stuffs, but it seems the goal is to have specific hw acceleration behavior.)